### PR TITLE
Introduce IWritableStorage interface

### PR DESCRIPTION
**Pull request by nupplaphil**
This PR is another preparation for additional, selectable storages.

I extracted all methods for selectable storage backends (currently Filesystem & DB) from `IStorage` to `ISelectableStorage`. So all the `BadMethodCallException`s aren't necessary anymore. I split the corresponding `get` operations of the StorageManager as well. So there's no need for an extra parameter `userBackend` anymore. Instead you use either the `getByName()` method to geht ANY backend, or you use the `getSelectableByName()` to get a specific, selectable backend. No need for any other checks.

Another improvement is the error handling. The backend per se doesn't need to know what the caller tries to do with the content. Therefore it shouldn't decide if an error of the persistence-layer is "acceptable" or not, it just throws an exception in case the call isn't right (e.g. you want to delete something, which doesn't exist, or get something, which doesn't exist). The caller is now responsible to catch the exception and do whatever he needs to do :).

Therefore I created two exceptions, the `ReferenceStorageExecption` for exceptions, where the caller uses invalid references and the `StorageException` for "the rest".

I used this branch (with Filesystem backend) for a couple of days without any exception so far.

### CODE REVIEW

**Comment by annando** [comment]
Would be great having a WebDAV storage.

File: doc/AddonStorageBackend.md [hunk 1]
```diff
@@ -106,21 +106,56 @@ Add a new test class which's naming convention is `StorageClassTest`, which exte
...
 } 
 ```
 
+## Exception handling
+
+There are two intended types of exceptions for storages
+
+### `StorageException`
+
+This is the common exception in case during the storage backend occurs any, unexpected errors.
```
Comment (MrPetovan) [inline doc/AddonStorageBackend.md]:
```suggestion
This is the common exception in case unexpected errors happen using the storage backend.
```

File: doc/AddonStorageBackend.md [hunk 2]
```diff
@@ -106,21 +106,56 @@ Add a new test class which's naming convention is `StorageClassTest`, which exte
...
+If there's a predecessor to this execption (e.g. you catched a exception an re-throw this execption), you should add the predecessor for transparent reasons.
```
Comment (MrPetovan) [inline doc/AddonStorageBackend.md]:
```suggestion
If there's a predecessor to this exception (e.g. you caught an exception and are throwing this execption), you should add the predecessor for transparency reasons.
```

File: doc/AddonStorageBackend.md [hunk 3]
```diff
@@ -106,21 +106,56 @@ Add a new test class which's naming convention is `StorageClassTest`, which exte
...
+		try {
+			throw new Exception('a real bad exception');
+		} catch (Exception $exception) {
+			throw new \Friendica\Model\Storage\StorageException(sprintf('The Example Storage throws an exception for reference %s', $reference), 500, $exception);
+		}
+	}
+} 
+```
+
+### `ReferenceStorageExecption`
```
Comment (MrPetovan) [inline doc/AddonStorageBackend.md]: Please move this more specific exception first.

File: doc/AddonStorageBackend.md [hunk 4]
```diff
@@ -106,21 +106,56 @@ Add a new test class which's naming convention is `StorageClassTest`, which exte
...
+This storage exception should be used in case the caller tries to use an invalid references.
+This could happen in case the caller tries to delete or update an unknown reference.
+The implementation of the storage backend must not ignore invalid references.
+
+Avoid throwing the common `StorageExecption` instead of the `ReferenceStorageException` at this particular situation!
+
 ## Example
 
-Here an hypotetical addon which register an unusefull storage backend.
+Here an hypotetical addon which register an unuseful storage backend.
```
Comment (MrPetovan) [inline doc/AddonStorageBackend.md]:
```suggestion
Here an hypotetical addon which register a useless storage backend.
```

**Comment by annando** [comment]
From my part it looks okay, but due to the complexity it would be great if @MrPetovan had got a look at it as well.

**Comment by MrPetovan** [comment]
I started looking at it in a car yesterday, but it was too big for my phone.

File: src/Model/Photo.php [hunk 5]
```diff
@@ -198,12 +201,17 @@ public static function getImageDataForPhoto(array $photo)
 			// legacy data storage in "data" column
 			$i = self::selectFirst(['data'], ['id' => $photo['id']]);
 			if ($i === false) {
-				return null;
+				return '';
```
Comment (MrPetovan) [inline src/Model/Photo.php]: Why the change of return value type?

File: src/Core/StorageManager.php [hunk 6]
```diff
@@ -73,65 +71,111 @@ class StorageManager
...
+	/**
+	 * Returns a selectable storage backend class by registered name
+	 *
+	 * @param string|null $name            Backend name
+	 *
+	 * @return Storage\ISelectableStorage|null null if no backend registered at $name
+	 *
+	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 */
+	public function getSelectableStorageByName(string $name = null)
```
Comment (MrPetovan) [inline src/Core/StorageManager.php]: Shouldn't this method throw an exception when the backend isn't valid?

File: src/Model/Storage/Database.php [hunk 7]
```diff
@@ -39,47 +38,57 @@ class Database extends AbstractStorage
 
 	/**
 	 * @param DBA             $dba
-	 * @param LoggerInterface $logger
-	 * @param L10n            $l10n
 	 */
-	public function __construct(DBA $dba, LoggerInterface $logger, L10n $l10n)
```
Comment (MrPetovan) [inline src/Model/Storage/Database.php:45]: Why did you forgo the logger and the l10n dependencies?

File: src/Model/Storage/Filesystem.php [hunk 8]
```diff
@@ -96,8 +97,7 @@ private function createFoldersForFile(string $file)
 
 		if (!is_dir($path)) {
 			if (!mkdir($path, 0770, true)) {
-				$this->logger->warning('Failed to create dir.', ['path' => $path]);
-				throw new StorageException($this->l10n->t('Filesystem storage failed to create "%s". Check you write permissions.', $path));
+				throw new StorageException(sprintf('Filesystem storage failed to create "%s". Check you write permissions.', $path));
```
Comment (MrPetovan) [inline src/Model/Storage/Filesystem.php:100]: Why didn't you keep the localized error messages?

File: src/Model/Storage/ISelectableStorage.php [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace Friendica\Model\Storage;
+
+/**
+ * Interface for selectable storage backends
```
Comment (MrPetovan) [inline src/Model/Storage/ISelectableStorage.php]: Can you please extend on what "selectable" means in this context?

File: src/Model/Storage/ReferenceStorageException.php [hunk 10]
```diff
@@ -21,31 +21,11 @@
...
-use Friendica\Core\L10n;
-use Psr\Log\LoggerInterface;
+use Exception;
 
 /**
- * A general storage class which loads common dependencies and implements common methods
+ * Storage Exception in case of invalid references
  */
-abstract class AbstractStorage implements IStorage
+class ReferenceStorageException extends Exception
```
Comment (MrPetovan) [inline src/Model/Storage/ReferenceStorageException.php]: Shouldn't it extend `StorageException`?

File: src/Model/Storage/ReferenceStorageException.php [hunk 11]
```diff
@@ -21,31 +21,11 @@
 
 namespace Friendica\Model\Storage;
```
Comment (MrPetovan) [inline src/Model/Storage/ReferenceStorageException.php:22]: Not sure these exception classes should be in the `Model\` namespace, maybe we should start a new namespace for all the Friendica exceptions.

File: src/Util/HTTPSignature.php [hunk 12]
```diff
@@ -408,7 +409,7 @@ public static function fetch($request, $uid)
 	 *                         'nobody' => only return the header
 	 *                         'cookiejar' => path to cookie jar file
 	 *
-	 * @return object CurlResult
+	 * @return CurlResult|void CurlResult
```
Comment (MrPetovan) [inline src/Util/HTTPSignature.php]: This shouldn't return `void` but throw an Exception if the `$owner` variable is empty, but maybe it's outside of the scope of this PR.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

**Comment by annando** [comment]
@nupplaphil have a look at the conflict.

File: src/Model/Storage/ReferenceStorageException.php (hunk 11 above)
Comment (nupplaphil) [reply src/Model/Storage/ReferenceStorageException.php:22]: I'd prefer having the exceptions at the place where the business logic resides. I think it's better to have explicit exceptions rather than "global" exceptions. f.e. I'm not a fan that we use the `InternalServerException` in cases of the backend errors. It's too common and should be the "Response" to a caller, not the "root-cause"

File: src/Model/Storage/Filesystem.php (hunk 8 above)
Comment (nupplaphil) [reply src/Model/Storage/Filesystem.php:100]: There's no need for localized errors at backend exceptions. They aren't meant to be printed at the frontend, so it could happen that there's a catch with a log-command and so you would get a (user-based) translated exception message inside your log :D

File: src/Model/Storage/Database.php (hunk 7 above)
Comment (nupplaphil) [reply src/Model/Storage/Database.php:45]: I thought the exceptions superseded the log-points. For me there's no need to call a log-point and throw the same exception to the caller.
l10n --> see https://github.com/friendica/friendica/pull/10582#discussion_r686248259

File: src/Model/Storage/ReferenceStorageException.php (hunk 10 above)
Comment (nupplaphil) [reply src/Model/Storage/ReferenceStorageException.php]: hmm ... good catch .. I wasn't sure if the `RefernceStorageException` is a child or a sibling ... but you're right, I guess child is better ...

File: src/Model/Storage/ReferenceStorageException.php (hunk 11 above)
Comment (MrPetovan) [reply src/Model/Storage/ReferenceStorageException.php:22]: Got it.

File: src/Model/Storage/Filesystem.php (hunk 8 above)
Comment (MrPetovan) [reply src/Model/Storage/Filesystem.php:100]: It could happen, but does it happen?

File: src/Model/Storage/ISelectableStorage.php (hunk 9 above)
Comment (nupplaphil) [reply src/Model/Storage/ISelectableStorage.php]: Maybe you have a better naming for this interface: 

It's a storage backend, which an admin can select for "user data" and have the CRUD (create update delete) pattern. The `IStorage` is a much simpler interface which allows just to retrieve data from a storage backend.

File: src/Model/Storage/Filesystem.php (hunk 8 above)
Comment (nupplaphil) [reply src/Model/Storage/Filesystem.php:100]: OK, but it's one of the rare places where we use translated exceptions .. But I don't see a use case for it .. what's the reason for translated exceptions? For me it just feels wrong tbh ^^ .. exceptions are "unwanted behaviors", mostly because of data/logic issues, addressed to admins and developers

File: doc/AddonStorageBackend.md (hunk 3 above)
Comment (nupplaphil) [reply doc/AddonStorageBackend.md]: done

File: src/Core/StorageManager.php (hunk 6 above)
Comment (nupplaphil) [reply src/Core/StorageManager.php]: Done

File: src/Model/Storage/ReferenceStorageException.php (hunk 10 above)
Comment (nupplaphil) [reply src/Model/Storage/ReferenceStorageException.php]: Done

File: src/Util/HTTPSignature.php (hunk 12 above)
Comment (nupplaphil) [reply src/Util/HTTPSignature.php]: I reverted it .. it's really part of another PR

The reason for it: https://github.com/friendica/friendica/blob/218d451248c585b9c97915d852d4bacbff838ecf/src/Util/HTTPSignature.php#L422 and https://github.com/friendica/friendica/blob/218d451248c585b9c97915d852d4bacbff838ecf/src/Util/HTTPSignature.php#L427

File: src/Model/Photo.php (hunk 5 above)
Comment (nupplaphil) [reply src/Model/Photo.php]: I think it's out of scope as well, but the callers are using the return value either as object .. or as string .. see:
https://github.com/friendica/friendica/blob/c6b929d6d932f1881e0697356efa702f00e2981b/src/Model/Photo.php#L230-L235
The first argument of `Image` assert that it's a string, not an object

I'll revert it, because I'm not sure if there're side effects ..

File: src/Core/StorageManager.php (hunk 6 above)
Comment (nupplaphil) [reply src/Core/StorageManager.php]: I reverted this change, because there are a lot of places (mainly inside the legacy api), where an "invalid" backend results in using the legacy storage method.

Comment (MrPetovan) [reply src/Core/StorageManager.php]: Can't this be done in a try-catch block instead of checking the returned value of the method?

File: src/Model/Storage/Filesystem.php (hunk 8 above)
Comment (MrPetovan) [reply src/Model/Storage/Filesystem.php:100]: But if it's triggered by a user, and we transmit the exception message to display, then it's important to show it in their preferred language. I understand that it wouldn't be of much help to them, but it still is more welcoming than error messages in a language you don't master.

File: src/Model/Storage/ISelectableStorage.php (hunk 9 above)
Comment (MrPetovan) [reply src/Model/Storage/ISelectableStorage.php]: Why not `Writable`?

File: src/Core/StorageManager.php (hunk 6 above)
Comment (nupplaphil) [reply src/Core/StorageManager.php]: You're right .. but I need to adapt all calls so I tried to avoid it :D

Comment (MrPetovan) [reply src/Core/StorageManager.php]: When a function main use is this:
```php
$return = function($param);
if ($return) {
   ...
} else {
   // Awkward error handling
}
```
Then it really should be
```php
try {
    $return = function($param);
    ...
} catch (Throwable $e) {
    // Proper error handling
}
```
The structures are similar and easy to transform from one to the other, it just is more explicit in the latter case.

File: src/Model/Storage/Filesystem.php (hunk 8 above)
Comment (nupplaphil) [reply src/Model/Storage/Filesystem.php:100]: currently we wouldn't display the exception to the user. There's a catch around it and a log-point *gg*. see:
https://github.com/friendica/friendica/blob/0c9163f299e693c853cbcecd05b88523ec9a1a3f/src/Model/Photo.php#L193-L217

But this behavior is now more straight, because before this change, the DB call returned null and the storage call threws an exception. I'll add a generic check at the `Module\Photo` and display the output to the user.

File: src/Model/Storage/ISelectableStorage.php (hunk 9 above)
Comment (nupplaphil) [reply src/Model/Storage/ISelectableStorage.php]: sounds better ^^

File: src/Model/Storage/Filesystem.php (hunk 8 above)
Comment (MrPetovan) [reply src/Model/Storage/Filesystem.php:100]: This works for me.

File: src/Core/StorageManager.php (hunk 6 above)
Comment (nupplaphil) [reply src/Core/StorageManager.php]: of course, you're right, but I need to take into account that in case of the "legacy" backend, there's no "current backend" set, so it would throw this exception all the time when calling `DI::storage()` ..

**Comment by nupplaphil** [comment]
OK, this is now A LOT bigger than I wanted .. I'm testing the current branch for a couple of days on my node ..

File: mod/photos.php [hunk 13]
```diff
@@ -323,7 +323,11 @@ function photos_post(App $a)
 			$photo = Photo::getPhotoForUser($page_owner_uid, $resource_id);
 
 			if (DBA::isResult($photo)) {
-				$image = Photo::getImageForPhoto($photo);
+				try {
+					$image = Photo::getImageForPhoto($photo);
+				} catch (Exception $exception) {
+					throw new HTTPException\InternalServerErrorException(DI::l10n()->t('Internal server error'), $exception);
+				}
```
Comment (MrPetovan) [inline mod/photos.php]: The `Image` class already has a built-in method to check the validity of the parameter, please add this try-catch block in `Model\Photo::getImageDataForPhoto()` and make `Model\Photo::getImageForPhoto()` return an `Object\Image` instance no matter what.

This saves you having to add this try-catch block everywhere since we were already using `Object\Image->isValid()`.

File: src/Core/StorageManager.php [hunk 14]
```diff
@@ -72,82 +69,135 @@ class StorageManager
...
 	 *
-	 * @return Storage\IStorage|null null if no backend registered at $name
+	 * @return Storage\IWritableStorage
 	 *
-	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 * @throws Storage\InvalidClassStorageException in case there's no backend class for the name
+	 * @throws Storage\StorageException in case of an unexpected failure during the hook call
 	 */
-	public function getByName(string $name = null, $onlyUserBackend = false)
+	public function getWritableStorageByName(string $name)
```
Comment (MrPetovan) [inline src/Core/StorageManager.php]: This is madness, you have two identical methods except one method call parameter changed from `true` to `false`. A better way to do this is to pass the array of valid backends to `getByName` like this:
```php
public function getWritableStorageByName(string $name)
{
    return $this->getByName($name, $this->backends);
}

public function getByName(string $name, $validBackends = null)
{
    ...
    if ($this->isValidBackend($name, $validBackends)) {
        ...
    }
    ...
}

public function isValidBackend(string $name = null, array $validBackends = null): bool
{
    $validBackends = $validBackends ?? array_merge($this->backends, [Storage\SystemResource::getName(), Storage\ExternalResource::getName()]);
    return in_array($name, $validBackends);
}
```

File: src/Model/Photo.php [hunk 15]
```diff
@@ -186,26 +189,29 @@ public static function exists(array $conditions)
...
 		if (!empty($photo['data'])) {
 			return $photo['data'];
 		}
 
-		$backendClass = DI::storageManager()->getByName($photo['backend-class'] ?? '');
-		if (empty($backendClass)) {
+		try {
+			$backendClass = DI::storageManager()->getByName($photo['backend-class'] ?? '');
+			$backendRef   = $photo['backend-ref'] ?? '';
+			return $backendClass->get($backendRef);
```
Comment (MrPetovan) [inline src/Model/Photo.php]: Unnecessary intermediate variable.
```suggestion
			return $backendClass->get($photo['backend-ref'] ?? '');
```

File: database.sql [hunk 16]
```diff
@@ -1,6 +1,6 @@
 -- ------------------------------------------
 -- Friendica 2021.09-dev (Siberian Iris)
--- DB_UPDATE_VERSION 1432
+-- DB_UPDATE_VERSION 1433
```
Comment (annando) [inline database.sql]: If my current PR #10593 is merged first, you will have to manually increate the database version. The system will not show a conflict here.

File: src/Model/Storage/Filesystem.php [hunk 17]
```diff
@@ -118,23 +118,34 @@ private function createFoldersForFile(string $file)
...
 	{
 		$file = $this->pathForRef($reference);
 		if (!is_file($file)) {
-			return '';
+			throw new ReferenceStorageException(sprintf('Filesystem storage failed to get the file %s, The file is invalid', $reference));
+		}
+
+		$result = file_get_contents($file);
+
+		// just in case the result is REALLY false, not zero or empty or anything else, throw the exception
```
Comment (MrPetovan) [inline src/Model/Storage/Filesystem.php]: This comment is entirely redundant with the code itself.
```suggestion
```

File: src/Model/Storage/IWritableStorage.php [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+ *
+ */
+
+namespace Friendica\Model\Storage;
+
+/**
+ * Interface for writable storage backends
+ *
+ * Used for storages with CRUD functionality, mainly used for user data (e.g. photos, attachements).
+ * There's only one active, writable storage possible. This type of storages are selectable by the current administrator
```
Comment (MrPetovan) [inline src/Model/Storage/IWritableStorage.php]: Did you have a free comma you had to put somewhere? 😄
```suggestion
 * There's only one active writable storage possible. This type of storage is selectable by the current administrator.
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/Core/StorageManager.php (hunk 14 above)
Comment (nupplaphil) [reply src/Core/StorageManager.php]: Looks a lot more readable, thx for this hint!

**Review by MrPetovan — APPROVED** [review]

**Comment by nupplaphil** [comment]
@MrPetovan I think I'm done , I will watch the error output until tomorrow and mark it ready for merge, but you already can take a look once more

File: database.sql (hunk 16 above)
Comment (nupplaphil) [reply database.sql]: addressed here: https://github.com/friendica/friendica/pull/10582/commits/92d532db6cac4a52840a003a3e62e3a4064bed51
thx for the hint!

File: src/Core/StorageManager.php [hunk 19]
```diff
@@ -46,7 +45,7 @@ class StorageManager
 		Storage\Database::NAME   => Storage\Database::class,
 	];
 
-	private $backends = [];
+	private $backends;
```
Comment (MrPetovan) [inline src/Core/StorageManager.php]: We should explicit this array as a list of IWritableStorage class names.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

**Review by MrPetovan — APPROVED** [review]
I asked for a one-line comment 😂

**Comment by nupplaphil** [comment]
seems to work .. ready for merge :)

File: src/Core/StorageManager.php [hunk 20]
```diff
@@ -61,93 +62,115 @@ class StorageManager
...
+	public function getWritableStorageByName(string $name): Storage\IWritableStorage
 	{
-		// @todo 2020.09 Remove this call after 2 releases
-		$name = $this->checkLegacyBackend($name);
+		$storage = $this->getByName($name, $this->validBackends);
+		if ($storage instanceof Storage\IWritableStorage) {
+			return $storage;
+		} else {
+			throw new Storage\InvalidClassStorageException(sprintf('Backend %s is not writable', $name));
+		}
```
Comment (MrPetovan) [inline src/Core/StorageManager.php]: Nitpicking, this structure can get rid of the `else`.
```suggestion
		if (!$storage instanceof Storage\IWritableStorage) {
			throw new Storage\InvalidClassStorageException(sprintf('Backend %s is not writable', $name));
		}
		
		return $storage;
```

File: src/Core/StorageManager.php [hunk 21]
```diff
@@ -61,93 +62,115 @@ class StorageManager
...
+	 * @throws Storage\StorageException in case of an unexpected failure during the hook call
+	 */
+	public function getByName(string $name, array $validBackends = null): Storage\IStorage
+	{
 		// If there's no cached instance create a new instance
 		if (!isset($this->backendInstances[$name])) {
 			// If the current name isn't a valid backend (or the SystemResource instance) create it
-			if ($this->isValidBackend($name, $onlyUserBackend)) {
+			if ($this->isValidBackend($name, $validBackends)) {
 				switch ($name) {
```
Comment (MrPetovan) [inline src/Core/StorageManager.php]: Same principle here with removing an indentation level.
```suggestion
			if (!$this->isValidBackend($name, $validBackends)) {
				throw new Storage\InvalidClassStorageException(sprintf('Backend %s is not valid', $name));
			}
			
			switch ($name) {
```

File: src/Core/StorageManager.php [hunk 22]
```diff
@@ -61,93 +62,115 @@ class StorageManager
...
-							$this->backendInstances[$data['name'] ?? $name] = $data['storage'];
-						} else {
-							return null;
+						try {
+							Hook::callAll('storage_instance', $data);
+							if (($data['storage'] ?? null) instanceof Storage\IStorage) {
+								$this->backendInstances[$data['name'] ?? $name] = $data['storage'];
+							} else {
+								throw new Storage\InvalidClassStorageException(sprintf('Backend %s was not found', $name));
+							}
```
Comment (MrPetovan) [inline src/Core/StorageManager.php]: More simplification
```suggestion
							if (!($data['storage'] ?? null) instanceof Storage\IStorage) {
								throw new Storage\InvalidClassStorageException(sprintf('Backend %s was not found', $name));
							}
							
							$this->backendInstances[$data['name'] ?? $name] = $data['storage'];
```

File: src/Core/StorageManager.php [hunk 23]
```diff
@@ -227,16 +231,20 @@ public function listBackends()
...
 	 */
-	public function register(string $class)
+	public function register(string $class): bool
 	{
 		if (is_subclass_of($class, Storage\IStorage::class)) {
 			/** @var Storage\IStorage $class */
 
-			$backends                    = $this->backends;
-			$backends[$class::getName()] = $class;
+			if (array_search($class::getName(), $this->validBackends, true) !== false) {
```
Comment (MrPetovan) [inline src/Core/StorageManager.php]: Look at this!
```suggestion
			if ($this->isValidBackend($class::getName(), $this->validBackends)) {
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/Core/StorageManager.php [hunk 24]
```diff
@@ -61,93 +62,118 @@ class StorageManager
...
 		// If there's no cached instance create a new instance
 		if (!isset($this->backendInstances[$name])) {
 			// If the current name isn't a valid backend (or the SystemResource instance) create it
-			if ($this->isValidBackend($name, $onlyUserBackend)) {
-				switch ($name) {
+			if (!$this->isValidBackend($name, $validBackends)) {
+				throw new Storage\InvalidClassStorageException(sprintf('Backend %s is not valid', $name));
+			}
+			
+			switch ($name) {
```
Comment (MrPetovan) [inline src/Core/StorageManager.php:142]: I wasn't able to suggest the full modification of the file, you have to deindent the rest of the `switch` cases and remove the final `else` yourself.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/Core/StorageManager.php (hunk 24 above)
Comment (nupplaphil) [reply src/Core/StorageManager.php:142]: I already saw it :-) .. I rebased the current branch onto the develop branch and did some "smoke tests" again (upload image, post image, set image as avatar) ... thx for the hint!

**Review by MrPetovan — APPROVED** [review]
